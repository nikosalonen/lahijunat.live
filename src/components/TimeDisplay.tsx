import { useMemo } from "preact/hooks";
import { useLanguageChange } from '../hooks/useLanguageChange';
import type { Train } from "../types";
import { t } from "../utils/translations";
import TimeRow from "./TimeRow";
// Memoize TimeDisplay
const TimeDisplay = ({
	departureRow,
	arrivalRow,
	timeDifferenceMinutes,
	isCancelled,
}: {
	departureRow: Train["timeTableRows"][0];
	arrivalRow?: Train["timeTableRows"][0];
	timeDifferenceMinutes: number;
	isCancelled?: boolean;
}) => {
	useLanguageChange();
	return useMemo(
		() => (
			<span class={`text-lg font-medium ${isCancelled ? 'line-through text-gray-500 dark:text-gray-400' : 'text-gray-800 dark:text-gray-200'} break-words min-w-0 relative pt-6`}>
				{departureRow.liveEstimateTime && timeDifferenceMinutes > 0 ? (
					<>
						<output
							aria-label={`${t('late')} ${timeDifferenceMinutes} ${t('minutes')}`}
							class="absolute top-0 left-0 px-2 py-0.5 bg-[#fed100] text-black rounded text-sm"
						>
							{`+${timeDifferenceMinutes} min`}
						</output>
						<TimeRow departureRow={departureRow} arrivalRow={arrivalRow} />
					</>
				) : (
					<TimeRow departureRow={departureRow} arrivalRow={arrivalRow} />
				)}
			</span>
		),
		[departureRow, arrivalRow, timeDifferenceMinutes, isCancelled],
	);
};

export default TimeDisplay;
