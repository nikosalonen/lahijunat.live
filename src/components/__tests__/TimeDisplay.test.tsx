import { render } from "@testing-library/preact";
import { describe, expect, it, vi } from "vitest";
import type { Train } from "../../types";
import TimeDisplay from "../TimeDisplay";

// Mock translations
vi.mock("../../utils/translations", () => ({
	t: (key: string) => {
		const translations: Record<string, string> = {
			late: "Myöhässä",
			minutes: "minuuttia",
			departure: "Lähtöaika",
			cancelled: "Peruttu",
		};
		return translations[key] || key;
	},
}));

// Mock useLanguageChange hook
vi.mock("../../hooks/useLanguageChange", () => ({
	useLanguageChange: vi.fn(),
}));

describe("TimeDisplay", () => {
	const mockDepartureRow: Train["timeTableRows"][0] = {
		stationShortCode: "HKI",
		type: "DEPARTURE",
		scheduledTime: "2024-03-20T10:00:00.000Z",
		liveEstimateTime: "2024-03-20T10:00:00.000Z",
		actualTime: "2024-03-20T10:00:00.000Z",
		differenceInMinutes: 0,
		trainStopping: true,
		commercialStop: true,
		commercialTrack: "1",
		cancelled: false,
		station: {
			name: "Helsinki",
			shortCode: "HKI",
			location: {
				latitude: 60.1699,
				longitude: 24.9384,
			},
		},
	};

	const mockArrivalRow: Train["timeTableRows"][0] = {
		stationShortCode: "TPE",
		type: "ARRIVAL",
		scheduledTime: "2024-03-20T11:00:00.000Z",
		liveEstimateTime: "2024-03-20T11:00:00.000Z",
		actualTime: "2024-03-20T11:00:00.000Z",
		differenceInMinutes: 0,
		trainStopping: true,
		commercialStop: true,
		commercialTrack: "2",
		cancelled: false,
		station: {
			name: "Tampere",
			shortCode: "TPE",
			location: {
				latitude: 61.4978,
				longitude: 23.7609,
			},
		},
	};

	it("renders on-time train correctly", () => {
		const { getByText } = render(
			<TimeDisplay
				departureRow={mockDepartureRow}
				arrivalRow={mockArrivalRow}
				timeDifferenceMinutes={0}
			/>,
		);

		// Should show scheduled departure and arrival without tildes
		getByText("12.00");
		getByText("13.00");
	});

	it("renders delayed train correctly", () => {
		const delayedDepartureRow = {
			...mockDepartureRow,
			liveEstimateTime: "2024-03-20T10:15:00.000Z",
			differenceInMinutes: 15,
		};

		const { getByText } = render(
			<TimeDisplay
				departureRow={delayedDepartureRow}
				arrivalRow={mockArrivalRow}
				timeDifferenceMinutes={15}
			/>,
		);

		expect(getByText("+15 min")).toBeInTheDocument();
	});

	it("renders cancelled train correctly", () => {
		const { container } = render(
			<TimeDisplay
				departureRow={mockDepartureRow}
				arrivalRow={mockArrivalRow}
				timeDifferenceMinutes={0}
				isCancelled={true}
			/>,
		);

		expect(container.firstChild).toHaveClass("line-through");
		expect(container.firstChild).toHaveClass("text-gray-500");
	});

	it("renders without arrival row", () => {
		const { container } = render(
			<TimeDisplay departureRow={mockDepartureRow} timeDifferenceMinutes={0} />,
		);

		expect(container).toMatchSnapshot();
	});

	it("does not show delay indicator when train is on time", () => {
		const { queryByText } = render(
			<TimeDisplay
				departureRow={mockDepartureRow}
				arrivalRow={mockArrivalRow}
				timeDifferenceMinutes={0}
			/>,
		);

		expect(queryByText("+0 min")).not.toBeInTheDocument();
	});

	it("does not show delay indicator when train is cancelled", () => {
		const delayedDepartureRow = {
			...mockDepartureRow,
			liveEstimateTime: "2024-03-20T10:15:00.000Z",
			differenceInMinutes: 15,
		};

		const { queryByText } = render(
			<TimeDisplay
				departureRow={delayedDepartureRow}
				arrivalRow={mockArrivalRow}
				timeDifferenceMinutes={15}
				isCancelled={true}
			/>,
		);

		// The delay indicator should not be visible for cancelled trains
		const delayIndicator = queryByText("+15 min");
		expect(delayIndicator).not.toBeInTheDocument();
	});
});
