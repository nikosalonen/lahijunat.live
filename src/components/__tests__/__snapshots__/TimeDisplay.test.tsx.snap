// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TimeDisplay > renders on-time train correctly 1`] = `
<div>
  <span
    class="text-xl sm:text-2xl font-medium text-gray-800 dark:text-gray-100 break-words min-w-0 relative pt-8 sm:pt-6"
  >
    <span
      class="whitespace-nowrap text-gray-600 dark:text-gray-300 text-lg sm:text-lg"
    >
       
      <time
        datetime="2024-03-20T10:00:00.000Z"
      >
        12.00
      </time>
      <span
        class="mx-2"
      >
        →
      </span>
      13.00
    </span>
  </span>
</div>
`;

exports[`TimeDisplay > renders without arrival row 1`] = `
<div>
  <span
    class="text-xl sm:text-2xl font-medium text-gray-800 dark:text-gray-100 break-words min-w-0 relative pt-8 sm:pt-6"
  >
    <span
      class="whitespace-nowrap text-gray-600 dark:text-gray-300 text-lg sm:text-lg"
    >
       
      <time
        datetime="2024-03-20T10:00:00.000Z"
      >
        12.00
      </time>
      <span
        class="mx-2"
      >
        →
      </span>
    </span>
  </span>
</div>
`;
